R/graded_condition.R

```diff
@@ -51,6 +65,67 @@ capture_errors <- function(expr, on_error = NULL) {
   )
 }
 
+
+sys_calls_most_helpful <- function() {
+  # borrowing ideas from rstudio/shiny/blob/2360bde1/R/conditions.R#L434-L448
+  calls <- sys.calls()
+  calls <- calls[-length(calls)] # not this function, obviously
+  callnames <- calls_name(calls)
+  
+  # We want to locate the last call in the stack that isn't just error handling
+  # infrastructure. The following functions are part of the stack when an error
+  # is captured and surfaced in gradethis, the hope is that the last call before
+  # the final block of these functions is reasonably informative.
+  hideable <- callnames %in% c(".handleSimpleError", "h", "on_error", "on_graded")
+  
+  # we also want to find the highest-level gradethis call, which probably
+  # provides the full context of the code generating the error
+  gradethis_pattern <- "^(gradethis:::?)?(grade_this|grade_result)"
+  idx_gradethis <- grep(gradethis_pattern, callnames)
+  
+  if (!length(idx_gradethis)) {
+    # next best guess, the function one level above the first function called
+    # in the function returned by grade_this()
+    idx_with_code_feedback <- which(callnames == "with_maybe_code_feedback")
+    if (length(idx_with_code_feedback)) {
+      idx_gradethis <- min(idx_with_code_feedback) - 1
+    }
+  }
+  
+  list(
+    first = if (length(idx_gradethis)) calls[[min(idx_gradethis)]],
+    last = calls[[max(which(!hideable))]]
+  )
+}
+
+calls_name <- function(calls) {
+  # borrowed from rstudio/shiny/blob/2360bde1/R/conditions.R#L64-L76
+  sapply(calls, function(call) {
```

Comment:
Let's change from a `sapply` to a `vapply(,character(1))`

Reply:
Sounds good! (That function is pretty much a direct copy from shiny, btw)

Reply:
a12126f1c39fc85aba0e6113e0886627138366a6